### PR TITLE
Remove native_functions.yaml dependency from some reduction operators

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64499
* __->__ #64173
* #64172
* #64171
* #64170

This one also required restructuring the code a bit to move the kernel
code into seperate files. So, I've mainly focused on CUDA which is
where the real build-time issues are.

Differential Revision: [D30728581](https://our.internmc.facebook.com/intern/diff/D30728581)

### CODE REVIEW

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30728581).

File: aten/src/ATen/core/Tensor.cpp [hunk 1]
```diff
@@ -15,6 +15,18 @@ TensorBase TensorBase::__dispatch_contiguous(c10::MemoryFormat memory_format) co
...
+  Tensor self(*this);
+  at::_ops::fill__Scalar::call(self, fill_value);
+  return *this;
+}
+
+const TensorBase& TensorBase::zero_() const {
+  Tensor self(*this);
+  at::_ops::zero_::call(self);
+  return *this;
+}
```
Comment (ezyang) [inline aten/src/ATen/core/Tensor.cpp:28]: Special dispensation, I see...

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30728581).

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30728581).

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30728581).

**Pull request closed without merging**